File: content/snippets/fuzzy-match-company-names/n1.test.py
import ast
import time
from pathlib import Path

import numpy as np
import pytest

from n0 import similarity
from n1 import build_index, match

# Un registre de la taille d'un petit annuaire professionnel.
REGISTER = [
    "Boulangerie Martin SARL",
    "Boulangerie Dupont",
    "Menuiserie Dubois SA",
    "Dubois Menuiserie",
    "Café de la Gare",
    "SNCF",
    "Société Nationale des Chemins de fer Français",
]
SNCF_DEVELOPPEE = REGISTER[-1]

INDEX = build_index(REGISTER)


def ranked(query):
    return dict(match(INDEX, query, top_k=len(REGISTER)))


# ---------------------------------------------------------------------------
# Point de rupture
# ---------------------------------------------------------------------------


def test_point_de_rupture_le_sigle_n_atteint_pas_les_trois_premiers_double_par_deux_societes_sans_rapport():
    """
    « la raison sociale développée de la SNCF n'atteint même pas les trois
    premiers résultats de sa propre abréviation, doublée par deux entreprises sans rapport ».
    """
    top = match(INDEX, "SNCF", top_k=4)
    assert [name for name, _ in top] == ["SNCF", "Menuiserie Dubois SA", "Boulangerie Martin SARL", SNCF_DEVELOPPEE]
    # Doublée par un score, pas par l'ordre du registre : les deux scores sont strictement plus hauts.
    assert top[1][1] > top[3][1] and top[2][1] > top[3][1] > 0.0
    assert round(top[3][1], 12) == 0.010394932704


def test_point_de_rupture_ponderer_repare_l_ordre_des_mots_et_classe_la_bonne_boulangerie():
    """Témoin du point de rupture : « Pondérer les fragments rares répare l'ordre des mots et classe la bonne boulangerie devant l'autre »."""
    assert match(INDEX, "MARTIN BOULANGERIE") == match(INDEX, "Boulangerie Martin")
    scores = ranked("Boulangerie Martin")
    assert scores["Boulangerie Martin SARL"] > scores["Boulangerie Dupont"]


# ---------------------------------------------------------------------------
# Autres affirmations du niveau
# ---------------------------------------------------------------------------


def test_trouve_la_bonne_societe_en_premier():
    name, score = match(INDEX, "Boulangerie Martin")[0]
    assert name == "Boulangerie Martin SARL"
    assert round(score, 12) == 0.883177974427  # le même nombre en JavaScript


def test_un_fragment_rare_pese_plus_qu_un_fragment_courant():
    """
    « « boulangerie » figure dans la moitié du registre et n'apprend presque rien ;
    « quiquengrogne » y figure une fois et tranche la question. »
    """
    register = ["Boulangerie Martin", "Boulangerie Dupont", "Boulangerie de la Quiquengrogne",
                "Boulangerie Petit", "Garage Lemoine", "Fleurs Roux"]
    index = build_index(register)
    vocabulary, idf = index["vectoriser"].vocabulary_, index["vectoriser"].idf_
    assert idf[vocabulary["quiq"]] > idf[vocabulary["boul"]]
    assert match(index, "Quiquengrogne", top_k=1)[0] == pytest.approx(("Boulangerie de la Quiquengrogne", 0.8013573), abs=1e-6)
    # « boulangerie » seul ne tranche rien : quatre boulangeries entre 0,47 et 0,67.
    assert [n for n, _ in match(index, "Boulangerie", top_k=4)] == [
        "Boulangerie Petit", "Boulangerie Martin", "Boulangerie Dupont", "Boulangerie de la Quiquengrogne"]


def test_jaro_winkler_traite_tous_les_caracteres_de_la_meme_facon():
    """« N0 n'en a pas d'équivalent » : pour N0, la boulangerie voisine passe au-dessus du seuil de 0,85."""
    assert similarity("Boulangerie Martin", "Boulangerie Dupont") > 0.85
    assert ranked("Boulangerie Martin")["Boulangerie Dupont"] < 0.6


def test_les_fragments_ne_chevauchent_jamais_deux_mots():
    """« `char_wb` keeps n-grams inside word boundaries, so a fragment never straddles two words. »"""
    assert not [g for g in INDEX["vectoriser"].vocabulary_ if " " in g.strip()]


def test_l_ordre_des_mots_ne_coute_rien_ici_au_contraire_de_n0():
    """« « Martin Dubois » et « Dubois Martin » se rencontrent donc […] au contraire de N0, où il coûte presque tout. »"""
    assert match(INDEX, "Martin Dubois", top_k=len(REGISTER)) == match(INDEX, "Dubois Martin", top_k=len(REGISTER))
    assert similarity("Martin Dubois", "Dubois Martin") < 0.5


def test_un_pluriel_et_une_inversion_a_la_fois():
    assert match(INDEX, "Menuiseries Dubois")[0][0] == "Dubois Menuiserie"


def test_des_fragments_de_deux_a_quatre_caracteres_survivent_a_une_faute_de_frappe():
    """Commentaire : « short enough to survive a typo somewhere else in the word »."""
    name, score = match(INDEX, "Boulangrie Martin")[0]
    assert name == "Boulangerie Martin SARL" and round(score, 12) == 0.818982690918


def test_les_lignes_sont_de_longueur_un_donc_le_cosinus_est_un_produit_scalaire():
    """« TfidfVectorizer returns rows of length one, so the cosine similarity is just the dot product. »"""
    norms = np.sqrt(np.asarray(INDEX["matrix"].multiply(INDEX["matrix"]).sum(axis=1)).ravel())
    assert norms == pytest.approx(np.ones(len(REGISTER)), abs=1e-12)
    assert all(ranked(name)[name] == pytest.approx(1.0, abs=1e-12) for name in REGISTER)


def test_une_recherche_dans_dix_mille_noms_est_un_produit_de_matrices():
    """« la recherche est un produit de matrices plutôt qu'une boucle sur toutes les paires : c'est ce qui rend un registre entier consultable »."""
    names = [f"Entreprise {i} {chr(97 + i % 26)}{chr(97 + (i // 26) % 26)}" for i in range(10_000)]
    index = build_index(names)
    start = time.perf_counter()
    top = match(index, names[4242])
    assert time.perf_counter() - start < 0.5
    assert top[0][0] == names[4242]


def test_les_egalites_reviennent_dans_l_ordre_du_registre():
    """« A stable sort, so two names with the same score always come back in register order. »"""
    register = ["Dupont SA", "Martin", "Dupont SA", "Garage", "Dupont SA"]
    index = build_index(register)
    assert [n for n, _ in match(index, "", top_k=5)] == register
    duplicates = match(index, "Dupont", top_k=3)
    assert [n for n, _ in duplicates] == ["Dupont SA"] * 3
    assert len({s for _, s in duplicates}) == 1


def test_un_nom_que_personne_n_a_ecrit_obtient_quand_meme_un_classement():
    name, score = match(INDEX, "Kwyjibo")[0]
    assert name in REGISTER and score > 0.0


def test_n1_est_deterministe():
    first = match(build_index(REGISTER), "Boulangerie Martin", top_k=7)
    assert all(match(build_index(REGISTER), "Boulangerie Martin", top_k=7) == first for _ in range(10))


def test_n1_s_appuie_sur_scikit_learn_et_numpy():
    """risks : vendor_lock library."""
    source = (Path(__file__).parent / "n1.py").read_text(encoding="utf-8")
    modules = set()
    for node in ast.walk(ast.parse(source)):
        if isinstance(node, ast.Import):
            modules |= {alias.name.split(".")[0] for alias in node.names}
        elif isinstance(node, ast.ImportFrom):
            modules.add(node.module.split(".")[0])
    assert modules == {"numpy", "sklearn"}


def test_valeurs_epinglees_pour_le_jumeau_javascript():
    """Le JavaScript dit faire « la même arithmétique que scikit-learn » : ces nombres sont affirmés des deux côtés."""
    assert round(ranked("Société Générale")[SNCF_DEVELOPPEE], 12) == 0.469061234241
    assert round(ranked("A B")["Café de la Gare"], 12) == 0.105066545124
    assert round(ranked("İstanbul Ltd")["Boulangerie Martin SARL"], 12) == 0.178184011082
    emoji = build_index(["🍞🥐 Boulangerie Martin", "Boulangerie Dupont", "𝔄𝔅 Conseil"])
    assert round(match(emoji, "𝔄𝔅", top_k=1)[0][1], 12) == 0.475128643566


# ---------------------------------------------------------------------------
# Cas de production
# ---------------------------------------------------------------------------


def test_production_une_requete_vide_ne_marque_rien():
    assert all(score == 0.0 for _, score in match(INDEX, "", top_k=len(REGISTER)))


def test_production_un_registre_vide_est_refuse_par_une_erreur():
    with pytest.raises(ValueError):
        build_index([])


def test_production_top_k_nul_ou_plus_grand_que_le_registre():
    assert match(INDEX, "Martin", top_k=0) == []
    assert len(match(INDEX, "Martin", top_k=100)) == len(REGISTER)


def test_production_accent_decompose_ou_absent_garde_la_bonne_reponse_en_tete():
    """N1 ne normalise pas les accents : le score baisse, le classement tient."""
    for query, expected in (("Café de la Gare", 0.887063254874), ("CAFE DE LA GARE", 0.884139130742)):
        name, score = match(INDEX, query)[0]
        assert name == "Café de la Gare" and round(score, 12) == expected


def test_production_une_requete_de_cent_ko_termine():
    start = time.perf_counter()
    match(INDEX, "boulangerie martin " * 5000)
    assert time.perf_counter() - start < 2

File: content/snippets/fuzzy-match-company-names/n1.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { similarity } from './n0.js';
import { buildIndex, match } from './n1.js';

// Un registre de la taille d'un petit annuaire professionnel.
const REGISTER = [
  'Boulangerie Martin SARL',
  'Boulangerie Dupont',
  'Menuiserie Dubois SA',
  'Dubois Menuiserie',
  'Café de la Gare',
  'SNCF',
  'Société Nationale des Chemins de fer Français',
];
const SNCF_DEVELOPPEE = REGISTER.at(-1);

const INDEX = buildIndex(REGISTER);
const ranked = (query, index = INDEX, size = REGISTER.length) => Object.fromEntries(match(index, query, size));
const round12 = (x) => Number(x.toFixed(12));

// ---------------------------------------------------------------------------
// Point de rupture
// ---------------------------------------------------------------------------

test("point de rupture : le sigle n'atteint pas les trois premiers, doublé par deux sociétés sans rapport", () => {
  const top = match(INDEX, 'SNCF', 4);
  assert.deepEqual(top.map(([name]) => name), ['SNCF', 'Menuiserie Dubois SA', 'Boulangerie Martin SARL', SNCF_DEVELOPPEE]);
  // Doublée par un score, pas par l'ordre du registre.
  assert.ok(top[1][1] > top[3][1] && top[2][1] > top[3][1] && top[3][1] > 0);
  assert.equal(round12(top[3][1]), 0.010394932704);
});

test("point de rupture : pondérer répare l'ordre des mots et classe la bonne boulangerie", () => {
  assert.deepEqual(match(INDEX, 'MARTIN BOULANGERIE'), match(INDEX, 'Boulangerie Martin'));
  const scores = ranked('Boulangerie Martin');
  assert.ok(scores['Boulangerie Martin SARL'] > scores['Boulangerie Dupont']);
});

// ---------------------------------------------------------------------------
// Autres affirmations du niveau
// ---------------------------------------------------------------------------

test('trouve la bonne société en premier', () => {
  const [name, score] = match(INDEX, 'Boulangerie Martin')[0];
  assert.equal(name, 'Boulangerie Martin SARL');
  assert.equal(round12(score), 0.883177974427); // le même nombre en Python
});

test("un fragment rare pèse plus qu'un fragment courant", () => {
  const register = ['Boulangerie Martin', 'Boulangerie Dupont', 'Boulangerie de la Quiquengrogne',
    'Boulangerie Petit', 'Garage Lemoine', 'Fleurs Roux'];
  const index = buildIndex(register);
  assert.ok(index.idf.get('quiq') > index.idf.get('boul'));
  const [name, score] = match(index, 'Quiquengrogne', 1)[0];
  assert.equal(name, 'Boulangerie de la Quiquengrogne');
  assert.ok(Math.abs(score - 0.8013573) < 1e-6);
  assert.deepEqual(match(index, 'Boulangerie', 4).map(([n]) => n),
    ['Boulangerie Petit', 'Boulangerie Martin', 'Boulangerie Dupont', 'Boulangerie de la Quiquengrogne']);
});

test('Jaro-Winkler traite tous les caractères de la même façon', () => {
  assert.ok(similarity('Boulangerie Martin', 'Boulangerie Dupont') > 0.85);
  assert.ok(ranked('Boulangerie Martin')['Boulangerie Dupont'] < 0.6);
});

test('les fragments ne chevauchent jamais deux mots', () => {
  assert.deepEqual([...INDEX.idf.keys()].filter((g) => g.trim().includes(' ')), []);
});

test("l'ordre des mots ne coûte rien ici, au contraire de N0", () => {
  assert.deepEqual(match(INDEX, 'Martin Dubois', REGISTER.length), match(INDEX, 'Dubois Martin', REGISTER.length));
  assert.ok(similarity('Martin Dubois', 'Dubois Martin') < 0.5);
});

test('un pluriel et une inversion à la fois', () => {
  assert.equal(match(INDEX, 'Menuiseries Dubois')[0][0], 'Dubois Menuiserie');
});

test('des fragments de deux à quatre caractères survivent à une faute de frappe', () => {
  const [name, score] = match(INDEX, 'Boulangrie Martin')[0];
  assert.equal(name, 'Boulangerie Martin SARL');
  assert.equal(round12(score), 0.818982690918);
});

test('les lignes sont ramenées à la longueur un, donc le cosinus est un produit scalaire', () => {
  for (const vector of INDEX.vectors) {
    assert.ok(Math.abs(Math.hypot(...vector.values()) - 1) < 1e-12);
  }
  for (const name of REGISTER) assert.ok(Math.abs(ranked(name)[name] - 1) < 1e-12, name);
});

test('une recherche dans dix mille noms termine vite', () => {
  const names = Array.from({ length: 10_000 }, (_, i) => `Entreprise ${i} ${String.fromCharCode(97 + (i % 26))}${String.fromCharCode(97 + (Math.floor(i / 26) % 26))}`);
  const index = buildIndex(names);
  const start = performance.now();
  const top = match(index, names[4242]);
  assert.ok(performance.now() - start < 500);
  assert.equal(top[0][0], names[4242]);
});

test("les égalités reviennent dans l'ordre du registre", () => {
  const register = ['Dupont SA', 'Martin', 'Dupont SA', 'Garage', 'Dupont SA'];
  const index = buildIndex(register);
  assert.deepEqual(match(index, '', 5).map(([n]) => n), register);
  const duplicates = match(index, 'Dupont', 3);
  assert.deepEqual(duplicates.map(([n]) => n), ['Dupont SA', 'Dupont SA', 'Dupont SA']);
  assert.equal(new Set(duplicates.map(([, s]) => s)).size, 1);
});

test("un nom que personne n'a écrit obtient quand même un classement", () => {
  const [name, score] = match(INDEX, 'Kwyjibo')[0];
  assert.ok(REGISTER.includes(name) && score > 0);
});

test('n1 est déterministe', () => {
  const first = match(buildIndex(REGISTER), 'Boulangerie Martin', 7);
  for (let i = 0; i < 10; i += 1) assert.deepEqual(match(buildIndex(REGISTER), 'Boulangerie Martin', 7), first);
});

test("n1 est écrit en entier, sans dépendance", () => {
  const source = readFileSync(new URL('./n1.js', import.meta.url), 'utf8');
  assert.doesNotMatch(source, /^\s*import\s|require\(/m);
});

test('la même arithmétique que scikit-learn : valeurs épinglées en Python', () => {
  assert.equal(round12(ranked('Société Générale')[SNCF_DEVELOPPEE]), 0.469061234241);
  assert.equal(round12(ranked('A B')['Café de la Gare']), 0.105066545124);
  assert.equal(round12(ranked('İstanbul Ltd')['Boulangerie Martin SARL']), 0.178184011082);
});

test('la même arithmétique que scikit-learn, hors du plan multilingue de base aussi', () => {
  // « Code points, as scikit-learn counts them » : valeurs épinglées en Python.
  const emoji = buildIndex(['🍞🥐 Boulangerie Martin', 'Boulangerie Dupont', '𝔄𝔅 Conseil']);
  assert.equal(round12(match(emoji, '𝔄𝔅', 1)[0][1]), 0.475128643566);
  assert.equal(round12(match(emoji, '🍞🥐 Boulangerie', 1)[0][1]), 0.76304806389);
});

test("le score s'explique fragment par fragment", () => {
  // Docstring N2 : « a score that cannot be explained fragment by fragment, as N1's can ».
  // Une requête identique à un nom du registre a le vecteur de ce nom.
  const [requete, nom] = [INDEX.vectors[1], INDEX.vectors[0]]; // « Boulangerie Dupont », « Boulangerie Martin SARL »
  const parts = new Map([...requete].filter(([g]) => nom.has(g)).map(([g, w]) => [g, w * nom.get(g)]));
  const somme = [...parts.values()].reduce((a, b) => a + b, 0);
  assert.ok(Math.abs(somme - ranked('Boulangerie Dupont')['Boulangerie Martin SARL']) < 1e-12);
  assert.equal(parts.size, 33);
  assert.ok([...parts.values()].every((p) => p > 0));
  // L'explication se lit : tout le score partagé vient du mot « boulangerie ».
  assert.ok([...parts.keys()].every((g) => ' boulangerie '.includes(g)));
});

// ---------------------------------------------------------------------------
// Cas de production
// ---------------------------------------------------------------------------

test('production : une requête vide ne marque rien', () => {
  assert.ok(match(INDEX, '', REGISTER.length).every(([, score]) => score === 0));
});

test('production : un registre vide rend une liste vide', () => {
  assert.deepEqual(match(buildIndex([]), 'Martin'), []);
});

test('production : topK nul ou plus grand que le registre', () => {
  assert.deepEqual(match(INDEX, 'Martin', 0), []);
  assert.equal(match(INDEX, 'Martin', 100).length, REGISTER.length);
});

test('production : accent décomposé ou absent garde la bonne réponse en tête', () => {
  for (const [query, expected] of [['Café de la Gare', 0.887063254874], ['CAFE DE LA GARE', 0.884139130742]]) {
    const [name, score] = match(INDEX, query)[0];
    assert.equal(name, 'Café de la Gare');
    assert.equal(round12(score), expected);
  }
});

test('production : une requête de cent Ko termine', () => {
  const start = performance.now();
  match(INDEX, 'boulangerie martin '.repeat(5000));
  assert.ok(performance.now() - start < 2000);
});
